# Task 6: Data Transformation & Feature Engineering

## Objective
Perform feature engineering on a housing dataset using Pandas and Scikit-learn.

## Tools Used
- Python
- Pandas
- Scikit-learn

## Techniques Used
- One-Hot Encoding
- Feature Scaling
- Interaction Terms

# Import Libraries

In [1]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import StandardScaler

# Loading Dataset

Loading the housing dataset into a Pandas DataFrame.

In [2]:
df = pd.read_csv("Housing.csv")

df.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


# Dataset Information

Checking dataset dimensions and data types.

In [3]:
print("Shape:", df.shape)

df.info()

Shape: (1460, 81)
<class 'pandas.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   str    
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   str    
 6   Alley          91 non-null     str    
 7   LotShape       1460 non-null   str    
 8   LandContour    1460 non-null   str    
 9   Utilities      1460 non-null   str    
 10  LotConfig      1460 non-null   str    
 11  LandSlope      1460 non-null   str    
 12  Neighborhood   1460 non-null   str    
 13  Condition1     1460 non-null   str    
 14  Condition2     1460 non-null   str    
 15  BldgType       1460 non-null   str    
 16  HouseStyle     1460 non-null   str    
 17  OverallQual    1460 non-null   int64  
 18  O

# Dataset Information

## Dataset Overview

The housing dataset contains 1460 records and 81 features.

### Data Types

- Integer Columns: 35
- Float Columns: 3
- Categorical Columns: 43

### Target Variable

The target variable is **SalePrice**, which represents the final sale price of a house.

### Observation

The dataset contains both numerical and categorical features, making it suitable for feature engineering techniques such as:

- Missing Value Handling
- One-Hot Encoding
- Feature Scaling
- Interaction Feature Creation

These transformations help improve the quality of data before machine learning modeling.

# Missing Values Analysis

Checking for missing values in the dataset.

In [4]:
df.isnull().sum().sort_values(ascending=False).head(10)

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
MasVnrType       872
FireplaceQu      690
LotFrontage      259
GarageQual        81
GarageFinish      81
GarageType        81
dtype: int64

# Missing Value Treatment

Before feature engineering, missing values must be handled to ensure data quality.

### Numerical Columns
Missing values are replaced using the median value.

### Categorical Columns
Missing values are replaced using the most frequent category (mode).

This approach preserves the dataset while minimizing the impact of missing data.

In [5]:
# Separate numerical and categorical columns

numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = df.select_dtypes(include=['object']).columns

# Fill numerical missing values with median

for col in numerical_cols:
    df[col] = df[col].fillna(df[col].median())

# Fill categorical missing values with mode

for col in categorical_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

# Check again

df.isnull().sum().sum()

C:\Users\DELL\AppData\Local\Temp\ipykernel_3396\75287930.py:4: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include=['object']).columns


np.int64(0)

# Missing Value Treatment Results

After applying median imputation for numerical features and mode imputation for categorical features, all missing values were successfully removed from the dataset.

This ensures that the dataset is complete and ready for feature engineering techniques such as One-Hot Encoding and Feature Scaling.

# One-Hot Encoding

Many machine learning algorithms cannot work directly with categorical variables.

One-Hot Encoding converts categorical features into numerical binary columns.

This transformation allows machine learning models to process categorical information effectively.

In [6]:
categorical_cols = df.select_dtypes(include=['object']).columns

print("Number of categorical columns:", len(categorical_cols))

Number of categorical columns: 43


C:\Users\DELL\AppData\Local\Temp\ipykernel_3396\3383669342.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include=['object']).columns


In [7]:
df_encoded = pd.get_dummies(
    df,
    columns=categorical_cols,
    drop_first=True
)

print("Original Shape:", df.shape)
print("Encoded Shape:", df_encoded.shape)

Original Shape: (1460, 81)
Encoded Shape: (1460, 246)


# One-Hot Encoding Results

After encoding, categorical features were transformed into binary variables.

This increased the number of columns in the dataset because each category was converted into a separate feature.

# Feature Scaling

Feature scaling is used to standardize numerical features.

StandardScaler transforms data so that:
- Mean = 0
- Standard Deviation = 1

This prevents features with larger values from dominating machine learning algorithms.

In [9]:
from sklearn.preprocessing import StandardScaler

numerical_cols = df_encoded.select_dtypes(
    include=['int64', 'float64']
).columns

scaler = StandardScaler()

df_encoded[numerical_cols] = scaler.fit_transform(
    df_encoded[numerical_cols]
)

df_encoded.head()

,Id,MSSubClass,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,BsmtFinSF1,...,SaleType_ConLI,SaleType_ConLw,SaleType_New,SaleType_Oth,SaleType_WD,SaleCondition_AdjLand,SaleCondition_Alloca,SaleCondition_Family,SaleCondition_Normal,SaleCondition_Partial
0,-1.730865,0.073375,-0.220875,-0.207142,0.651479,-0.517200,1.050994,0.878668,0.514104,0.575425,...,False,False,False,False,True,False,False,False,True,False
1,-1.728492,-0.872563,0.460320,-0.091886,-0.071836,2.179628,0.156734,-0.429577,-0.570750,1.171992,...,False,False,False,False,True,False,False,False,True,False
2,-1.726120,0.073375,-0.084636,0.073480,0.651479,-0.517200,0.984752,0.830215,0.325915,0.092907,...,False,False,False,False,True,False,False,False,True,False
3,-1.723747,0.309859,-0.447940,-0.096897,0.651479,-0.517200,-1.863632,-0.720298,-0.570750,-0.499274,...,False,False,False,False,True,False,False,False,False,False
4,-1.721374,0.073375,0.641972,0.375148,1.374795,-0.517200,0.951632,0.733308,1.366489,0.463568,...,False,False,False,False,True,False,False,False,True,False


# Feature Scaling Results

Numerical features were standardized using StandardScaler.

The transformed features now have a consistent scale, which improves model performance and training stability.

# Creating Interaction Features

Interaction features are generated by combining existing features.

These features help machine learning models capture relationships between variables that may not be apparent when using individual features alone.

In [10]:
df_encoded["Quality_Area_Interaction"] = (
    df["OverallQual"] * df["GrLivArea"]
)

df_encoded[
    [
        "OverallQual",
        "GrLivArea",
        "Quality_Area_Interaction"
    ]
].head()

,OverallQual,GrLivArea,Quality_Area_Interaction
0,0.651479,0.370333,11970
1,-0.071836,-0.482512,7572
2,0.651479,0.515013,12502
3,0.651479,0.383659,12019
4,1.374795,1.299326,17584


# Interaction Feature Results

A new feature named Quality_Area_Interaction was created by multiplying Overall Quality and Ground Living Area.

This feature represents the combined effect of house quality and living area on property value.

# Engineered Dataset Preview

The dataset has now been transformed through:

- Missing Value Treatment
- One-Hot Encoding
- Feature Scaling
- Interaction Feature Engineering

The resulting dataset is ready for machine learning applications.

In [11]:
df_encoded.head()

,Id,MSSubClass,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,BsmtFinSF1,...,SaleType_ConLw,SaleType_New,SaleType_Oth,SaleType_WD,SaleCondition_AdjLand,SaleCondition_Alloca,SaleCondition_Family,SaleCondition_Normal,SaleCondition_Partial,Quality_Area_Interaction
0,-1.730865,0.073375,-0.220875,-0.207142,0.651479,-0.517200,1.050994,0.878668,0.514104,0.575425,...,False,False,False,True,False,False,False,True,False,11970
1,-1.728492,-0.872563,0.460320,-0.091886,-0.071836,2.179628,0.156734,-0.429577,-0.570750,1.171992,...,False,False,False,True,False,False,False,True,False,7572
2,-1.726120,0.073375,-0.084636,0.073480,0.651479,-0.517200,0.984752,0.830215,0.325915,0.092907,...,False,False,False,True,False,False,False,True,False,12502
3,-1.723747,0.309859,-0.447940,-0.096897,0.651479,-0.517200,-1.863632,-0.720298,-0.570750,-0.499274,...,False,False,False,True,False,False,False,False,False,12019
4,-1.721374,0.073375,0.641972,0.375148,1.374795,-0.517200,0.951632,0.733308,1.366489,0.463568,...,False,False,False,True,False,False,False,True,False,17584


# Saving Engineered Dataset

The transformed dataset is saved for future machine learning modeling and analysis.

In [12]:
df_encoded.to_csv(
    "engineered_housing_dataset.csv",
    index=False
)

print("Engineered dataset saved successfully!")

Engineered dataset saved successfully!


# Conclusion

In this project, feature engineering techniques were applied to a housing dataset.

## Transformations Performed

1. Missing Value Treatment
2. One-Hot Encoding
3. Feature Scaling using StandardScaler
4. Interaction Feature Creation

## Outcome

The dataset was successfully transformed into a machine-learning-ready format.

These preprocessing techniques improve data quality, enhance feature representation, and contribute to better predictive model performance.